src/openfermion/transforms/_binary_code_transform.py

```diff
@@ -0,0 +1,187 @@
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+""" The transform function that does Fermion-qubit mappings
+    based on a BinaryCode (arXiv:1712.07067) """
+
+import numpy
+
+from openfermion.ops import (BinaryCode,
+                             FermionOperator,
+                             QubitOperator,
+                             SymbolicBinary)
+
+
+def _extract(term):
+    """Extractor superoperator helper function.
+
+    Args:
+        term (tuple): a single summand from a binary expression
+
+    Returns (QubitOperator or int): the qubit operator that corresponds
+        to the binary expression or -1
+    """
+    if len(term) == 1:
+        term = term[0]
+        if term[1] == 'W':
+            return QubitOperator('Z' + str(term[0]))
+        elif term[1] == '1':
+            return -1
+    if len(term) > 1:
+        return dissolve(term)
+
+
+def extractor(binary_op):
+    """ Applies the extraction superoperator to a binary expression
+     to obtain the corresponding qubit operators.
+
+    Args:
+        binary_op (SymbolicBinary): the binary term
+
+    Returns (QubitOperator): the qubit operator corresponding to the
+        binary terms
+    """
+    return_fn = 1
+    for term in binary_op.terms:
+        return_fn *= _extract(term)
+    return return_fn
+
+
+def dissolve(term):
+    """Decomposition helper. Takes a product of binary variables
+    and outputs the Pauli-string sum that corresponds to the
+    decomposed multi-qubit operator.
+
+    Args:
+        term (tuple): product of binary variables, i.e.: 'w0 w2 w3'
+
+    Returns (QubitOperator): superposition of Pauli-strings
+
+    Raises:
+        ValueError: if the action is not 'W'
```

Comment:
Why "W" where does this convention come from exactly? I am concerned that this design is rather opaque.

Reply:
I'm not sure that is a very satisfying answer for you, but 'w' is just labeling the binary variables. It stems from the SymbolicBinary class.  At some point writing the paper I had to decide for symbols for the binary vectors of the Fermion modes and qubit configuration. I just called them \nu and \omega. We don't need  \nu here, but \omega  survived as 'w'.

Reply:
So I am starting to question whether the data structure for SymbolicBinary is implemented in a reasonable fashion. What other value can the action take, just the empty string. In this case, it seems to me that the whole data structure is overcomplicated. Instead of storing  (1, 'W'), (2, 'W') shouldn't you just store (1,2)? To represent constants you can use "none", or something like that.

Reply:
I actually experimented with different ways of defining the SymbolicBinary class - even tried the (1,2,..) version which would indeed simplify things a lot for the operations - couple of problems I had with that was the constant definition and user input. 
for constant indeed a string is a good idea, did not think of that. As for user definitions would we not allow for str inputs? would we stick to 'W' in str inputs - if not what if the user gives 1+q1+w2+z3 is that ok? or do we assume that anyone thats using the symbolicBinary class knows what they are doing?

Reply:
It's definitely more streamlined this way, you are right. We basically made it like it is now just because appears not to be that different from the data structures of FermionOperator and QubitOperator. But it's your call:  would you like us to change (1, 'W'), (2, 'W') into (1,2) and so on?

Reply:
I guess also when i started writing this I was wondering if it could adopt from SymbolicOperator class as well.

Reply:
Yes, I think this is a good change to make. I would actually suggest "None", for a constant, but you could put a string. It's very easy to write "if action is None:" and it is probably less arbitrary than any string you could choose. I can tell you were thinking of inheriting from SybolicOperator, but since you didn't, I think the change makes sense.

Regarding string initialization, I think that is still fine. You can probably allow them to use any char in fact. So if they wrote "a1 a2 + a3 a4" or "W1 W3 + W3 W4" it would do the same thing. Another option is to represent the term a1 a2 with (1 2) in the string input also. So you could write (1 2) + (3 4). I personally find the second option cleaner but that is up to you.

Another thing to think about is whether you want to enforce numerical order. For instance, if you want to disallow people from having a term (3,1) as opposed to (1,3). If you do this, you can certainly perform certain numerical operations faster since you know the list is sorted. But unless numerical performance is a bottleneck it is not necessary to specialize the magic methods for this now (but it might still he a good idea to put systems in place that prevent having both (1 3) and (3 1).

Sorry I did not catch this before!

Reply:
Ok, we'll make the change. We already take care of several mishaps during the input, like a messed up order or additions with constants other than one, and so on.

Reply:
Awesome, thanks!

Reply:
I'm ok with it all but None. its a dangerous name to use for a variable. certain data structures are built to ignore None's by default - and defining 1 as None is confusing to me.

Reply:
That's fine. If you feel a string would be better that seems reasonable.